### ISSUE TITLE
Last readable stream example crashes

### ISSUE THREAD

**Issue by starikovs** [substack/stream-handbook#90]
If I run `printf "Hello, World!" | node program.js` on the last reading stream example I get an error:

    /vagrant/program.js:4
        var b = process.stdin.read();
                       ^
    
    RangeError: Maximum call stack size exceeded
    ...

Here's the example:

    var offset = 0;
    
    process.stdin.on('readable', function () {
        var buf = process.stdin.read();
        if (!buf) return;
        for (; offset < buf.length; offset++) {
            if (buf[offset] === 0x0a) {
                console.dir(buf.slice(0, offset).toString());
                buf = buf.slice(offset + 1);
                offset = 0;
                process.stdin.unshift(buf);
                return;
            }
        }
        process.stdin.unshift(buf);
    });

But if I run `echo` instead of `printf` there's not an error. I think it's because of the last `process.stdin.unshift(buf);` and I don't understand what its purpose.. Can anybody help me? Thanks in advance!

**Comment by noffle** [comment]
Good catch! This is happening because the function keeps on reading until it sees a newline. However, `printf` defaults to no newline. You'd see the *intended* behaviour if you had of run `printf 'heya\n' | node stream.js`.

What's happening underneath is that `process.stdin.unshift(buf)` causes `"readable"` to be emitted. This happens over and over again, growing the call stack until Node crashes (since no more newlines are ever coming).

In practice you'd probably do something like [`split`](https://github.com/dominictarr/split/blob/master/index.js#L54) and buffer pieces elsewhere to avoid the recursion. This lets you detect when the incoming stream has ended and gracefully abort.

---

I think a good edit to the handbook would be to mention that this causes the stack to grow and prevents the stream's close|end events from being fired (since the unshift makes it seem like there's still content).
